gbpservice/neutron/services/servicechain/plugins/ncp/node_drivers/oneconvergence_service_node_driver.py

```diff
@@ -0,0 +1,867 @@
+
+        is_template_aws_version = stack_template.get(
+                                        'AWSTemplateFormatVersion', False)
+        resources_key = ('Resources' if is_template_aws_version
+                         else 'resources')
+        parameters_key = ('Parameters' if is_template_aws_version
+                          else 'parameters')
+        properties_key = ('Properties' if is_template_aws_version
+                          else 'properties')
+
+        insert_type = 'north_south' if is_consumer_external else 'east_west'
+
+        if service_type == pconst.LOADBALANCER:
+            self._generate_pool_members(context, stack_template,
+                                        config_param_values,
+                                        provider_ptg,
+                                        is_template_aws_version)
+
+        # copying to _plugin_context should not be required if we are not
+        # mixing service chain context with plugin context anywhere
+        admin_context = self._get_admin_context()
+        service_info = self.svc_mgr.get_service_info_with_srvc_type(
+                context=context.plugin_context, service_type=service_type,
+                tenant_id=context.plugin_context.tenant_id,
+                insert_type=insert_type)
+
+        # If we are going to share an already launched VM, we do not have
+        # to create new ports/PTs for management and stitching
+        if service_info:
+            floating_ip = service_info['floating_ip']
+            provider_port_id = service_info['provider_port_id']
+            provider_port_mac = context.core_plugin.get_port(
+                    admin_context, provider_port_id)['mac_address']
+        else:
+            raise ServiceInfoNotAvailableOnUpdate()
+
+        if service_type != pconst.LOADBALANCER:
+            if service_type == pconst.FIREWALL:
+                if not is_consumer_external:
+                    consumer_ptg_subnet_id = context.consumer['subnets'][0]
+                    consumer_subnet = context.core_plugin.get_subnet(
+                        context._plugin_context, consumer_ptg_subnet_id)
+                    consumer_cidr = consumer_subnet['cidr']
+                else:
+                    consumer_cidr = '0.0.0.0/0'
+
+                provider_cidr = provider_subnet['cidr']
+                stack_template = self._update_template_with_firewall_rules(
+                    context, provider_ptg, provider_cidr, consumer_cidr,
+                    stack_template, is_template_aws_version)
+                firewall_desc = {'vm_management_ip': floating_ip,
+                                 'provider_ptg_info': [provider_port_mac],
+                                 'insert_type': insert_type}
+                stack_template[resources_key]['Firewall'][properties_key][
+                    'description'] = str(firewall_desc)
+            elif service_type == pconst.VPN:
```

Comment:
at present we don't support VPN/Firewall in L2 mode. But for future support, should we check for insertion mode in all alls that relate to stritching.